src/Features/Core/Portable/SplitOrMergeIfStatements/AbstractSplitIfStatementCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractSplitIfStatementCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        protected abstract int GetLogicalExpressionKind(ISyntaxKindsService syntaxKinds);
+
+        protected abstract CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument, string ifKeywordText);
+
+        protected abstract Task<SyntaxNode> GetChangedRootAsync(
+            Document document,
+            SyntaxNode root,
+            SyntaxNode ifLikeStatement,
+            SyntaxNode leftCondition,
+            SyntaxNode rightCondition,
+            CancellationToken cancellationToken);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(context.Span.Start);
+
+            if (context.Span.Length > 0 &&
+                context.Span != token.Span)
+            {
+                return;
+            }
+
+            var syntaxFacts = context.Document.GetLanguageService<ISyntaxFactsService>();
+            var syntaxKinds = context.Document.GetLanguageService<ISyntaxKindsService>();
+            var ifGenerator = context.Document.GetLanguageService<IIfLikeStatementGenerator>();
+
+            if (IsPartOfBinaryExpressionChain(token, GetLogicalExpressionKind(syntaxKinds), out var rootExpression) &&
```

Comment:
fwiw, i find "syntaxKinds" confusing here.  It reads liek this is an `ImmutableArray<SyntaxKind>`.

Reply:
testing

Reply:
Considering that it's declared just 3 lines above and the type is apparent, I'd prefer to leave it as is.